californium-core/src/test/java/org/eclipse/californium/core/test/MessageExchangeStoreTool.java

```diff
@@ -237,9 +248,46 @@ public void onTimeout() {
 		public boolean allRequestsTerminated() {
 			return requests.isEmpty();
 		}
-		
-		public Collection<Request> getUnterminatedRequests(){
+
+		public Collection<Request> getUnterminatedRequests() {
 			return requests;
 		}
 	}
+
+	public static class TestEndpointContextMatcher extends UdpEndpointContextMatcher {
+
+		/**
+		 * Ignore address. Plain coap is not supporting observe/notify with
+		 * changed addresses. To make such tests, please use
+		 * {@link CoapTestEndpoint#CoapTestEndpoint(InetSocketAddress, NetworkConfig, boolean)}
+		 */
+		private final boolean ignoreAddress;
+
+		/**
+		 * Create context matcher.
+		 * 
+		 * @param ignoreAddress {@code true}, if address should be ignored,
+		 *            {@code false}, otherwise.
+		 */
+		public TestEndpointContextMatcher(boolean ignoreAddress) {
+			this.ignoreAddress = ignoreAddress;
+		}
+
+		/**
+		 * {@inheritDoc}
+		 * 
+		 * Depending on {@link #ignoreAddress}, the address is ignored for
+		 * matching.
+		 */
+		@Override
+		public boolean isResponseRelatedToRequest(EndpointContext requestContext, EndpointContext responseContext) {
+			if (ignoreAddress) {
+				return isToBeSent(requestContext, responseContext);
```

Comment:
Just a detail : I would use `internalMatch` instead of `isToBeSent`. (and so change internalMatch to protected)

This make me read again `internalMatch` and
 -  I still have difficulty to understand the `inhibitNewConnection`
 - is it possible to have an availableContext == null ?

Reply:
inhibitNewConnection is generally for cases, where a special EndpointContext is passed in (e.g. with DTLS session) but currently no "connection" is associated and therefore the connectionContext is null. 
For the UDPConnector this seems to be not too relevant, but you will get an error, if you use a EndpointContext which inhibitNewConnection returns true (e.g. with DTLS session).

Reply:
The main use case of `inhibitNewConnection` is here right ?
```java
DTLSSession session = connection.getEstablishedSession();
if (session == null) {
	if (!checkOutboundEndpointContext(message, null)) {
		return;
	}
	// no session with peer established yet, create new empty session &
	// start handshake
```

But I still not understand what is the concrete use case of this :/

So this is the case where you send a message and you want to explicitly forbid a new connection by passing a requestedEndpointContext with inhibitNewConnection ? The message will be send only if this is the same connection ?

Reply:
If you provide a DTLS Session ID in the EndpointContext, but you don't have already a session for that target, depending on the EndpointContextMatcher, it is useless to start an handshake, because this could only end in a different EndpointContext! 

The PrincipalEndpointContextMatcher doesn't use the inhibitNewConnection, because there it's possible to match even with a new handshake.
The Matchers derived from KeySetEndpointContextMatcher will use inhibitNewConnection, because here a new handshake could not end with the same session.

Reply:
So this is a kind of optimization ? and this is strongly dependent of contextMatcher policy ?
Maybe `inhibitNewConnection` is more linked to contextMatcher than to endpointContext itself ? Could it make sense to move `inhibitNewConnection` in context matcher instead ?

Reply:
It depends on the data contained in the EndpointContext and its usage in the EndpointContextMatcher. Therefore the inhibitNewConnection reflects the data, and the Matcher, if its used.

Reply:
What I mean : this seems more understandable to me if we have somethink like this:

`EndpointContextMatcher.canCreateNewConnection(EndpointContext requestedContext)`